File: TwoPointer/15. 3Sum/SortedTwoSum.java
class SortedTwoSum {
    public List<List<Integer>> threeSum(int[] nums) {
 
        Arrays.sort(nums);
        List<List<Integer>> ans = new ArrayList<>();
        for (int i = 0; i < nums.length; i++) {
           // no values to the right can sum to 0
          if (i == 0 || nums[i-1] != nums[i]) {
            ans.addAll(twoPointerSum(nums, i));
          }
        }

        return ans;
    }

    private List<List<Integer>> twoPointerSum(int[] nums, int i) {
      int low = i+1, high = nums.length -1;
      List<List<Integer>> ans = new ArrayList<>();

      while (low < high) {
        int val = nums[i] + nums[low] + nums[high];
        
        if (val < 0) {
          low++;
        } else if (val > 0) {
          high--;
        } else {
          ans.add(List.of(nums[i], nums[low++], nums[high--]));
          while (low < high && nums[low-1] == nums[low]) {
            ++low;
          }
        }
      }

      return ans;
    } 
}

File: TwoPointer/15. 3Sum/SortedTwoSum.py
class SortedTwoSum(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """

        nums.sort();
        res = []
        for i in range(len(nums)):
          if (i == 0 or nums[i-1] != nums[i]):
            self.twoSum(nums, i, res)
        return res

    def twoSum(self, nums, i, res):
      lo, hi = i+1, len(nums)-1
      while (lo < hi):
        sum = nums[i] + nums[lo] + nums[hi]
        if sum < 0:
            lo+=1
        elif sum > 0:
            hi-=1
        else:
            res.append([nums[i], nums[lo], nums[hi]])
            lo += 1
            hi -= 1
            while lo < hi and nums[lo] == nums[lo-1]:
                lo+=1
